### PR TITLE
First PR

### PR DESCRIPTION
**Pull request by benjamin99**
#### Desc
The PR setup the basic structure of the **Dcard-Gondola** service.
The service contains two parts, which are
1) the api service: to provide the RESTful interfaces for the admin panel or other services to *a)* get the status of specific member(s), *b)* list the status of the members in the waiting list, or *c)* promote the specified members in the waiting list to the second stage.
2) the stream consumer: to fetch the scope update events from the `Dcard-API` and put the corresponding member to the waiting list if the added scopes contains `graduate:queued`.

#### What had been done
1. the corresponding `service.proto` and `stream.proto`
2. the basic implementations for both `stream` and `service`
3. the basic implementations for the `dao` and the `dcard client`

#### TODOs
1. refine the implementations and complete the related test cases
2. CI/CD related

### CODE REVIEW

File: pkg/dcard/http_client.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+type HttpClient struct {
+	dcard.Client
+}
+
+type scopeRequest struct {
+	scopes []string
+}
+
+func (client *HttpClient) AddScopes(opt *ScopeOption) error {
+	endpoint := fmt.Sprintf("/v2/members/%d/sopces", opt.MemberID)
```
Comment (tommy351) [inline pkg/dcard/http_client.go]:
```suggestion
	endpoint := fmt.Sprintf("/v2/members/%d/scopes", opt.MemberID)
```

Comment (benjamin99) [reply pkg/dcard/http_client.go]: done

File: pkg/service/base.go [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+	return &proto.ShowListResponse{List: list}, nil
+}
+
+func (s *BaseService) Promote(ctx context.Context, req *proto.PromotionRequest) (*proto.PromotionResponse, error) {
+	opt := &dcard.ScopeOption{MemberID: 100, Scopes: []string{"graduate:accepted"}}
+	if err := s.AddScopes(opt); err != nil {
+		return nil, err
+	}
+
+	if err := s.statusDAO.Promote(100); err != nil { // TODO: should handle the case when the member had been marked promoted
```
Comment (rueian) [inline pkg/service/base.go]: 這上面兩個 100 member id 是打錯的？

File: pkg/streamproto/genkit.stream.transport_amqp.go [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+				var err error
+				ctx1, span := trace.StartSpan(ctx, "gondolastream.amqp.HandleEvent", trace.WithSpanKind(trace.SpanKindServer))
+				defer func() {
+					setSpanStatus(span, err)
+					span.End()
+				}()
+
+				req := &Event{}
+				if strings.Contains(msg.ContentType(), "application/json") || strings.Contains(msg.ContentType(), "text/plain") {
+					err = json.Unmarshal(msg.Data(), req)
```
Comment (rueian) [inline pkg/streamproto/genkit.stream.transport_amqp.go]: 現在 dcard-api-event 是送 json data 嗎？不過它應該沒有設定 content type 吧？

gen 出來的 amqp transport 在這邊可能會有問題

Comment (benjamin99) [reply pkg/streamproto/genkit.stream.transport_amqp.go]: event 現在是送 json data 沒錯

Comment (benjamin99) [reply pkg/streamproto/genkit.stream.transport_amqp.go]: https://github.com/Dcard/Dcard-API/pull/927 有關 content-type 我 dcard-api 開 PR 了

File: pkg/dcard/http_client.go [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+
+	"github.com/Dcard/Dcard-Go-SDK"
+)
+
+type HttpClient struct {
+	dcard.Client
+}
+
+type scopeRequest struct {
+	scopes []string
```
Comment (tommy351) [inline pkg/dcard/http_client.go]: private 欄位不會出現在 json 裡

```suggestion
	Scopes []string `json:"scopes"`
```

**Pull request merged**